File: app/src/main/java/com/ternaryop/utils/drawer/adapter/DrawerItem.java
package com.ternaryop.utils.drawer.adapter;

import android.app.Fragment;
import android.content.Context;

import com.ternaryop.utils.drawer.counter.CountRetriever;

public class DrawerItem {
    public final static DrawerItemDivider DRAWER_ITEM_DIVIDER = new DrawerItemDivider();
    private int itemId;
    private String title;
    private Class<? extends Fragment> fragmentClass;
    private boolean counterVisible;
    private CountRetriever countRetriever;
    private final boolean isHeader;
    private final boolean isDivider;
    private String badge;

    public DrawerItem() {
        this(-1, null, null, false);
    }

    public DrawerItem(int itemId, String title) {
        this(itemId, title, null, false);
    }
    
    public DrawerItem(int itemId, String title, Class<? extends Fragment> fragmentClass) {
        this(itemId, title, fragmentClass, false);
    }

    public DrawerItem(int itemId, String title, Class<? extends Fragment> fragmentClass, boolean showCounter) {
        this(title, fragmentClass, showCounter, null);
        this.itemId = itemId;
    }

    @Deprecated
    public DrawerItem(String title, Class<? extends Fragment> fragmentClass, boolean showCounter, CountRetriever countRetriever) {
        this.title = title;
        this.fragmentClass = fragmentClass;
        this.counterVisible = showCounter;
        this.countRetriever = countRetriever;
        this.isHeader = fragmentClass == null && title != null;
        this.isDivider = fragmentClass == null && title == null;
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public Class<? extends Fragment> getFragmentClass() {
        return fragmentClass;
    }

    public void setFragmentClass(Class<? extends Fragment> fragmentClass) {
        this.fragmentClass = fragmentClass;
    }

    public boolean isHeader() {
        return isHeader;
    }

    public boolean isDivider() {
        return isDivider;
    }

    public boolean isCounterVisible() {
        return counterVisible;
    }

    public void setCounterVisible(boolean showCounter) {
        this.counterVisible = showCounter;
    }

    public CountRetriever getCountRetriever() {
        return countRetriever;
    }

    public void setCountRetriever(CountRetriever countRetriever) {
        this.countRetriever = countRetriever;
    }

    public Fragment instantiateFragment(Context context) {
        return Fragment.instantiate(context, getFragmentClass().getName());
    }

    public String getBadge() {
        return badge;
    }

    public void setBadge(String badge) {
        this.badge = badge;
    }

    public int getItemId() {
        return itemId;
    }

    public void setItemId(int itemId) {
        this.itemId = itemId;
    }

    public static class DrawerItemDivider extends DrawerItem {
        @Override
        public void setTitle(String title) {
        }

        @Override
        public void setFragmentClass(Class<? extends Fragment> fragmentClass) {
        }

        @Override
        public void setCounterVisible(boolean showCounter) {
        }

        @Override
        public void setCountRetriever(CountRetriever countRetriever) {
        }

        @Override
        public void setBadge(String badge) {
        }

        @Override
        public void setItemId(int itemId) {
        }
    }
}

File: app/src/main/java/com/ternaryop/utils/drawer/adapter/DrawerItem.kt
package com.ternaryop.utils.drawer.adapter

import android.content.Context
import android.os.Bundle
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager

open class DrawerItem constructor(open var itemId: Int = -1,
    open var title: String? = null,
    open var fragmentClass: Class<out Fragment>? = null,
    open var isCounterVisible: Boolean = false,
    open var arguments: Bundle? = null,
    open var argumentsBuilder: (() -> Bundle?)? = null) {
    val isHeader: Boolean
        get() =  fragmentClass == null && title != null
    val isDivider: Boolean
        get() = fragmentClass == null && title == null
    open var badge: String? = null

    open fun instantiateFragment(context: Context, fragmentManager: FragmentManager): Fragment {
        val fragment = fragmentManager.fragmentFactory.instantiate(context.classLoader, fragmentClass!!.name)
        fragment.arguments = argumentsBuilder?.let { it() } ?: arguments

        return fragment
    }

    @Suppress("UNUSED_PARAMETER")
    class DrawerItemDivider : DrawerItem() {
        override var title: String?
            get() = super.title
            set(title) {}
        override var fragmentClass: Class<out Fragment>?
            get() = super.fragmentClass
            set(fragmentClass) {}
        override var isCounterVisible: Boolean
            get() = super.isCounterVisible
            set(showCounter) {}
        override var badge: String?
            get() = super.badge
            set(badge) {}
        override var itemId: Int
            get() = super.itemId
            set(itemId) {}
    }

    companion object {
        val DRAWER_ITEM_DIVIDER = DrawerItemDivider()
    }
}